allennlp_models/rc/models/transformer_qa.py

```diff
@@ -65,131 +78,187 @@ def forward(  # type: ignore
+        self,
+        best_spans: torch.Tensor,
+        span_start_logits: torch.Tensor,
+        span_end_logits: torch.Tensor,
+        answer_span: torch.Tensor,
+    ) -> torch.Tensor:
+        """
+        Calculate the loss against the `answer_span` and also update the  span metrics.
+        """
+        span_start = answer_span[:, 0]
+        span_end = answer_span[:, 1]
+        span_mask = span_start != -1
+        self._span_accuracy(best_spans, answer_span, span_mask.unsqueeze(-1).expand_as(best_spans))
 
-            self._span_start_accuracy(span_start_logits, span_start, span_mask)
-            self._span_end_accuracy(span_end_logits, span_end, span_mask)
+        start_loss = cross_entropy(span_start_logits, span_start, ignore_index=-1)
+        big_constant = min(torch.finfo(start_loss.dtype).max, 1e9)
+        assert not torch.any(start_loss > big_constant), "Start loss too high"
 
-            output_dict["loss"] = loss
+        end_loss = cross_entropy(span_end_logits, span_end, ignore_index=-1)
+        assert not torch.any(end_loss > big_constant), "End loss too high"
 
-        # Compute the EM and F1 on SQuAD and add the tokenized input to the output.
-        if metadata is not None:
-            best_spans = best_spans.detach().cpu().numpy()
+        self._span_start_accuracy(span_start_logits, span_start, span_mask)
+        self._span_end_accuracy(span_end_logits, span_end, span_mask)
 
-            output_dict["best_span_str"] = []
-            context_tokens = []
-            for metadata_entry, best_span, cspan in zip(metadata, best_spans, context_span):
-                context_tokens_for_question = metadata_entry["context_tokens"]
-                context_tokens.append(context_tokens_for_question)
+        return (start_loss + end_loss) / 2
 
+    def _collect_best_span_strings(
+        self,
+        best_spans: torch.Tensor,
+        context_span: torch.IntTensor,
+        metadata: List[Dict[str, Any]],
+        cls_index: Optional[torch.LongTensor],
+    ) -> List[str]:
+        """
+        Collect the string of the best predicted span from the context metadata and
+        update `self._per_instance_metrics`, which in the case of SQuAD v1.1 / v2.0
+        includes the EM and F1 score.
+        """
+        best_spans = best_spans.detach().cpu().numpy()
+
+        best_span_strings = []
+        for i, (metadata_entry, best_span, cspan) in enumerate(
+            zip(metadata, best_spans, context_span)
+        ):
+            context_tokens_for_question = metadata_entry["context_tokens"]
+
+            if cls_index is not None and best_span[0] == cls_index[i]:
+                # Predicting [CLS] is interpreted as predicting the question as unanswerable.
+                best_span_string = ""
+                # NOTE: even though we've "detached" 'best_spans' above, this still
+                # modifies the original tensor in-place.
+                best_span[0], best_span[1] = -1, -1
```

Comment:
Do we want that to happen?

Reply:
Yes because `best_spans` (from each `best_span`) are added to the output dict. And it's expected that this span is either a valid index into the context, or something else. And I think the "something else" makes sense to be `(-1, -1)`.

Reply:
Modifying the input parameters is not usually the way we return information. We can break the rules in some cases, but when we do we should slap a big warning on the function that this is what's happening.